crytic/slither
# Rename tuple assignments

### Notes

Consider the following code:
```
pragma solidity 0.8.13;

contract Test
{
    function f1() public view returns (string memory, int)  {
        return ("hello", 0);
    }

    function f2(bool a) public view returns (string memory)  {
        if (a) {
            (string memory q, int z) = f1();
            return q;
        } else {
            (string memory q, int z) = f1();
            return q;
        }

    }
}
```

When running slither on the above code, we get an incorrect uninitialized variable finding:
```
Test.f2(bool).q_scope_0 (source/test.sol#15) is a local variable never initialized
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#uninitialized-local-variables
```

This happens because solidity's AST+CFG generator renames variable definitions to avoid duplicate names. In the above code, both `q` and `z` are renamed in the else branch. However, `q` is renamed only at its occurrence in the return statement and not in its declaration. This happens because the declaration id is used to look up revised names, and the AST+CFG generator does not record the declaration id of the identifier in the declaration itself. (The id is in the json ast, but it is not copied to the python AST.) 

This change copies the declaration id into the `id` field of the ast. It also modifies the renaming code to look up renamings using the "id" field as well as the "referencedDeclaration" field. We could have copied the `id` field of the declaration identifier to the `referencedDeclaration` field, but it seems preferable to keep the json ast and the python ast as similar as possible.

### Testing
* Clone `slither-task`
* Copy the above code snippet to `./source/test.solc` under the `slither-task` root
* Clone this branch to `slither-task/utilities/slither`
* Run the above code snippet using `SOLC_VERSION=0.8.13 pipenv run slither ./source/test.solc` and verify there is no spurious uninitialized variable finding
* Diff the files in `output/commit/tmp/slither-out.json` generated with and without this branch's changes. Verify there is no difference.
* Run `./evaluate.sh run 100` and then `./evaluate.sh diff`. Verify that only difference is a single project that builds successfully now and failed previously.

### Related Issue
https://github.com/CertiKProject/slither-task/issues/191
